seaduck/smart_read.py

```diff
@@ -1,84 +1,98 @@
+        start = indexes.min() or None
+        stop = stop if (stop := indexes.max() + 1) < size else None
+        slicers += (slice(start, stop),)
+    indexes_tuple = tuple(
+        indexes.ravel() - slicer.start if slicer.start else indexes.ravel()
+        for indexes, slicer in zip(indexes_tuple, slicers)
+    )
+    return da.data[slicers], indexes_tuple
+
+
+def smart_read(da, indexes_tuple, dask_more_efficient=100, chunks="auto"):
+    """Read from a xarray.DataArray given a tuple indexes.
 
     Try to do it fast and smartly.
+    There is a lot of improvement to be made here,
+    but this is how it is currently done.
+
+    The data we read is going to be unstructured but they tend to be
+    rather localized. For example, the lagrangian particles read data
+    from the same time step.
+    This function figures out which chunks stores the data, convert them
+    into numpy arrays, and then read the data from the converted ones.
 
     Parameters
     ----------
     da: xarray.DataArray
         DataArray to read from
-    ind: tuple of numpy.ndarray
+    indexes_tuple: tuple of numpy.ndarray
         The indexes of points of interest, each element does not need to be 1D
-    memory_chunk: int, default 3
-        If the number of chunks needed is smaller than this, read all of them at once.
-    xarray_more_efficient: int, default 100
+    dask_more_efficient: int, default 100
         When the number of chunks is larger than this, and the data points are few,
-        it may make sense to directly use xarray's vectorized read.
+        it may make sense to directly use dask's vectorized read.
+    chunks: int, str, default: "auto"
+        Chunks for indexes
 
     Returns
     -------
     + values: numpy.ndarray
-        The values of the points of interest. Has the same shape as the elements in ind.
+        The values of the points of interest. Has the same shape as the elements in indexes_tuple.
     """
-    the_shape = ind[0].shape
-    ind = tuple(i.ravel() for i in ind)
-    if len(da.dims) != len(ind):
-        raise ValueError("index does not match the number of dimensions")
-    if da.chunks is None or da.chunks == {}:
-        npck = np.array(da)
-        return npck[ind].reshape(the_shape)
-    if (
-        np.prod([len(i) for i in da.chunks]) <= memory_chunk
-    ):  # if the number of chunks is small don't bother
-        npck = np.array(da)
-        return npck[ind].reshape(the_shape)
-    cksz = orderdic(zip(da.dims, da.chunks))
-    keys = list(cksz.keys())
-    n = len(ind[0])
-    result = np.zeros(n)
+    if len(indexes_tuple) != da.ndim:
+        raise ValueError(
+            "indexes_tuple does not match the number of dimensions: "
+            f"{len(indexes_tuple)} vs {da.ndim}"
+        )
+
+    shape = indexes_tuple[0].shape
+    size = indexes_tuple[0].size
+    if not size:
+        return np.empty(shape)  # TODO: Why shape (0, ...) is allowed and tested?
```

Comment:
@MaceKuailv why do you need to pass and return indexes with shape `(0, ...)`?
A few tests fail if I don't do this.

Reply:
Exactly which tests failed? There shouldn't be any place this is called with array that has shape (0,...). Are you talking about single numbers?

Reply:
```
FAILED tests/test_oceinterp.py::test_diff_oceinterp[ecco] - ValueError: zero-size array to reduction operation minimum which has no identity
FAILED tests/test_oceinterp.py::test_eulerian_oceinterp[varList2-ecco-x0-y0-z0-t0] - ValueError: zero-size array to reduction operation minimum which has no identity
FAILED tests/test_oceinterp.py::test_eulerian_oceinterp[varList3-ecco-x0-y0-z0-t0] - ValueError: zero-size array to reduction operation minimum which has no identity
```

Reply:
I've printed the input shape:
```
FAILED tests/test_oceinterp.py::test_diff_oceinterp[ecco] - AssertionError: (0, 1, 2)
FAILED tests/test_oceinterp.py::test_eulerian_oceinterp[varList2-ecco-x0-y0-z0-t0] - AssertionError: (0, 1, 1)
FAILED tests/test_oceinterp.py::test_eulerian_oceinterp[varList3-ecco-x0-y0-z0-t0] - AssertionError: (0, 1, 1)
```

Reply:
Still not totally clear to me. I'll remove this line a look from my side. Could it be because you remove the line that checks whether there are chunks?

Reply:
Sometimes for LLC grid you need v velocity to read u velocity. When all the points are away from the face connections, a read is initiated with shape zero and does not expect any value. 

Is this a very bad thing to do? I can make it a special case in the interpolation code.

Reply:
Doesn't look great, but could be OK if it's done on purpose and there's no better implementation.
The shapes printed above are just `indexes_tuple[0].shape`, so it's the original shape of the indexes.